rust/src/csv/reader.rs

```diff
@@ -149,21 +151,33 @@ fn infer_schema(mut file: File, delimiter: u8) -> Schema {
         .delimiter(delimiter)
         .from_reader(BufReader::new(file.try_clone().unwrap()));
 
-    // take at most 100 rows for schema inference
-    // [TODO] it's often better to read the whole file and infer its schema
-    for result in csv_reader.into_records().take(100) {
-        let record = result.unwrap();
+    match max_records {
+        Some(n) => {
+            for result in csv_reader.into_records().take(n) {
```

Comment:
Instead of repeating the logic, you can just get a single iterator based on max_records like this:

```
    let records: Box<Iterator<Item = Result<StringRecord>>> = match max_records {
        Some(n) => Box::new(csv.record_iter.take(n)),
        None => Box::new(csv.record_iter)
    };

    for record  in records {

    }
```

Reply:
I had tried reusing the logic, but couldn't get it to work. 

I'm not getting your suggestion to work,

```rust
error[E0271]: type mismatch resolving `<csv_crate::StringRecordsIntoIter<std::io::BufReader<std::fs::File>> as std::iter::Iterator>::Item == std::result::Result<csv_crate::StringRecord, error::ArrowError>`
   --> src\csv\reader.rs:155:63
    |
155 |       let records: Box<Iterator<Item = Result<StringRecord>>> = match max_records {
    |  _______________________________________________________________^
156 | |         Some(n) => Box::new(csv_reader.into_records().take(n)),
157 | |         None => Box::new(csv_reader.into_records())
158 | |     };
    | |_____^ expected struct `csv_crate::Error`, found enum `error::ArrowError`
    |
    = note: expected type `std::result::Result<_, csv_crate::Error>`
               found type `std::result::Result<_, error::ArrowError>`
    = note: required for the cast to the object type `dyn std::iter::Iterator<Item=std::result::Result<csv_crate::StringRecord, error::ArrowError>>
```

Reply:
OK not to worry I can make this minor optimization as a separate PR once this one is merged

Reply:
Yes, the code in this section does seem a little repetitive...